readme.txt
=== WM Simple Captcha ===
Contributors: http://plugins.web-mumbai.com/
Tags: simple captcha, registration captcha, custom captcha image, register captcha image, theme my captcha
Requires at least: 3.6
Tested up to: 4.2.2
Stable tag: 2.0
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Captcha image for registration page, customize according to your theme.

== Description ==

WM Simple Captcha for registration page, customize captcha image according to your theme. Change image width, height, background color, set possible characters, add space between two characters, adjust font size, change form label, add your custom css, enable/disable refresh button, change refresh button. Many think more on customization.

= Features =

* Enable/Disable Captcha for wordpress registration page
* Set the captcha image with
* Set  the captcha image height
* Set the captcha possible characters length
* Adjust font size
* Enable space between characters
* Change font
* Set possible letters
* Set random Dots
* Set random Line
* Set text color
* Set dots color
* Set line color
* Change form label
* Enable captcha border
* Set captcha border color
* Enable/Disable refresh button
* Change refresh images as your theme
* Not case sensitive

In Short you can change captcha image style as your theme style. 


== Installation ==

= Manual installation =

1.	Download the plugin file to your computer and unzip it
2.	Using an FTP program, or your hosting control panel, upload ` wm_simple_captcha `  folder to the `/wp-content/plugins/` directory
3.	Login to your WordPress admin panel. 
4.	Activate the plugin through the 'Plugins' menu in WordPress
5.	Go To setting -> wm_simple_captcha change default setting according to your theme.
6.	Plugin is ready for work.

= Automatic installation =

Automatic installation is the easiest option as WordPress handles the file transfers itself and you don't even need to leave your web browser. To do an automatic install, 

1.	Login to your WordPress admin panel.
2.	Navigate to the Plugins menu and click Add New.
3.	In the search field type " WM Simple Captcha" and click Search Plugins. 
4.	Once you've found the plugin you can view details about it such as the the point release, rating and description. 
5.	Most importantly of course, you can install it by clicking _Install Now_.
6.	Activate the plugin through the 'Plugins' menu in WordPress
7.	Go To setting -> wm_simple_captcha change default setting according to your theme.
8.	Plugin is ready for work.

== Screenshots ==

1. The overview screen registration.
2. "WM Simple Captcha setting" page.
3. Captcha Image Settings.
4. Form Settings.
5. Form Error and custom CSS

== Frequently Asked Questions ==

= Any detail documentation available? =
Yes, [Web Mumbai Plugins](http://plugins.web-mumbai.com/ "Web Mumbai Plugins") coming soon detail documentation.

= If enable/disable registration, I loose my "WM Simple Captcha" =
No, change your setting of captcha image, then enable this for registration page.

= Can I change background color? =
Yes, see the plugin settings page.

= Can I change captcha text color? =
Yes, see the plugin settings page.


== Upgrade Notice ==
= 2.0.0 = 2015-07-12
* Tested : Wordpress 4.2.2
* Added: Font Upload Functionality

= 1.1 = 2015-03-02
* Added : added 'uninstall.php' in 'wm-simple-captcha' root folder. When using 'uninstall.php' the plugin should always check for the WP_UNINSTALL_PLUGIN constant, before executing. The WP_UNINSTALL_PLUGIN constant is defined by WordPress at runtime during a plugin uninstall, it will not be present if 'uninstall.php' is requested directly.
* Removed : register_uninstall_hook().
* Bugfix : We fixed error on activation.

= 1.0 = 2015-03-01
* Added: First working release of the Plugin

== Changelog ==
= 2.0.0 = 2015-07-12
* Tested: tested in latest wordpress 4.2.2
* Added: Font Upload Functionality
		
= 1.1.1 = 2015-04-18
* Bugfix : We fixed error on activation.
* Tested : Wordpress 3.8.1
* Tested : Wordpress 3.8.2
* Tested : Wordpress 3.8.3
* Tested : Wordpress 3.9

= 1.1 = 2015-03-02
* Bugfix : We fixed error on activation.

= 1.0 = 2015-03-01
* Initial Version



changelog.txt
= 2.0.0 = 2015-07-12
* Tested: tested in latest wordpress 4.2.2
* Added: Font Upload Functionality
		
= 1.1.1 = 2015-04-18
* Bugfix : We fixed error on activation.
* Tested : Wordpress 3.8.1
* Tested : Wordpress 3.8.2
* Tested : Wordpress 3.8.3
* Tested : Wordpress 3.9

= 1.1 = 2015-03-02
* Bugfix : We fixed error on activation.

= 1.0 = 2015-03-01
* Initial Version


captcha_code/captcha_code.php
<?php 
session_start();
header('Content-Type: image/jpeg');// defining the image type to be shown in browser widow

//Settings: You can customize the captcha here
if(isset($_SESSION['wmsc_options'])){
	$wmsc_options 				= $_SESSION['wmsc_options'];
	$image_width 				= $wmsc_options['captcha_image_width'];
	$image_height 				= $wmsc_options['captcha_image_height'];
	$characters_on_image 		= $wmsc_options['captcha_image_characters'];
	$captcha_enable_space 		= isset_value('captcha_enable_space',$wmsc_options,0);
	$captcha_image_font_adj		= isset_value('captcha_image_font_adj',$wmsc_options,0.6);
	$font 						= $wmsc_options['captcha_image_font'];
	$possible_letters			= $wmsc_options['captcha_possible_letters'];
	$random_dots 				= $wmsc_options['captcha_random_dots'];
	$random_lines 				= $wmsc_options['captcha_random_lines'];
	$captcha_text_color			= "0x".$wmsc_options['captcha_text_color'];
	$captcha_dots_color			= "0x".$wmsc_options['captcha_dots_color'];
	$captcha_line_color 		= "0x".$wmsc_options['captcha_line_color'];
	$captcha_background_color	= "0x".$wmsc_options['captcha_background_color'];
	$font_path					= $wmsc_options['font_path'];
	
	
}else{
	$image_width 				= 120;
	$image_height 				= 25;
	$characters_on_image 		= 5;
	$font 						= 'arial.ttf';
	$possible_letters 			= '23456789bcdfghjkmnpqrstvwxyz';//avoid confusing characters (l 1 and i for example)
	$random_dots 				= 0;
	$random_lines 				= 0;
	$captcha_text_color			= "0x142864";
	$captcha_dots_color 		= "0x142864";
	$captcha_line_color 		= "0x142864";
	$captcha_background_color 	= "0xFFFFFF";
	$captcha_enable_space		= 0;
	$captcha_image_font_adj		= "0.6";
	$font_path					= "../fonts/";
}

$code 			= '';
$code_string 	= '';
$space 			= '';
$font 			= $font_path.$font;

if(!file_exists($font)) $font = "../fonts/arial.ttf";

if($captcha_enable_space == 1) $space = ' ';

$i = 0;
while ($i < $characters_on_image) {
	$c = substr($possible_letters, mt_rand(0, strlen($possible_letters)-1), 1);
	$code_string .= $c.$space;
	$code .= $c;
	$i++;
}

//$code_string = "Refresh";

$font_size 		= $image_height * $captcha_image_font_adj;
$image 			= @imagecreate($image_width, $image_height);


/* setting the background, text and dots colours here */
$arr_background_color = hexrgb($captcha_background_color);
$background_color = imagecolorallocate($image, $arr_background_color['red'], $arr_background_color['green'], $arr_background_color['blue']);

$arr_text_color = hexrgb($captcha_text_color);
$image_text_color = imagecolorallocate($image, $arr_text_color['red'], $arr_text_color['green'], $arr_text_color['blue']);

$arr_dots_color = hexrgb($captcha_dots_color);
$image_dots_color = imagecolorallocate($image, $arr_dots_color['red'], $arr_dots_color['green'], $arr_dots_color['blue']);

$arr_line_color = hexrgb($captcha_line_color);
$image_line_color = imagecolorallocate($image, $arr_line_color['red'], $arr_line_color['green'], $arr_line_color['blue']);


/* generating lines randomly in background of image */
for( $i=0; $i<$random_lines; $i++ ) {
	imageline($image, mt_rand(0,$image_width), mt_rand(0,$image_height),mt_rand(0,$image_width), mt_rand(0,$image_height), $image_line_color);
}

/* generating the dots randomly in background */
for( $i=0; $i<$random_dots; $i++ ) {
	imagefilledellipse($image, mt_rand(0,$image_width),	mt_rand(0,$image_height), 2, 3, $image_dots_color);
}


//$code_string = "WM Simple Captcha";

/* create a text box and add 6 letters code in it */
$textbox = imagettfbbox($font_size, 0, $font, $code_string); 
$x = ($image_width - $textbox[4])/2;
$y = ($image_height - $textbox[5])/2;
imagettftext($image, $font_size, 0, $x, $y, $image_text_color, $font , $code_string);



/* Show captcha image in the page html page */
imagejpeg($image);//showing the image
imagedestroy($image);//destroying the image instance
$_SESSION['6_letters_code'] = $code;

function isset_value ($name,$data = NULL, $default = ''){
	if(isset($data[$name]))	return $data[$name];
	return $default;
}

function hexrgb ($hexstr){
  $int = hexdec($hexstr);
  return array("red" => 0xFF & ($int >> 0x10),"green" => 0xFF & ($int >> 0x8),"blue" => 0xFF & $int);
}
?>

classes/wm_simple_captcha_front.php
<?php
session_start();

if ( ! defined( 'ABSPATH' ) ) exit; // Exit if accessed directly

if ( ! class_exists( 'WM_Simple_Captcha_Front' ) ) {
	class WM_Simple_Captcha_Front{		
		
		var $options_page = NULL;
		
		public function __construct($filename) {
			global $wmsc_options;
			
			$this->define_constant($filename);
			//delete_option('wmsimplecaptcha');
			$wmsc_options 	= $this->default_values();			
			$_SESSION['wmsc_options'] = $wmsc_options;
			
			if (is_admin()) {	
					$this->include_admin();
					$this->settings = new WM_Simple_Captcha_Admin($filename);
					register_deactivation_hook($filename, 	array(&$this, 'deactivate'));	// deactivation functions
					//register_activation_hook($filename, 	array(&$this, 'activate'));		// Activation functions
					//register_uninstall_hook($filename, 		array('WM_Simple_Captcha_Front', 'uninstall') ); 	// Register an uninstall hook to automatically remove options
			}else{
				if($wmsc_options['captcha_enable_registration']){
					//$this->print_array($_SESSION['wmsc_options']);
					add_action( 'register_form', 			array(&$this, 'wmcaptcha_display'),100,1);		
					add_action( 'registration_errors', 		array(&$this, 'registration_captcha_errors'),20,1);	
					add_action( 'init', 					array(&$this, 'wmsimplecaptcha_scripts_front'),20,1);
					
					
					//For login screen
					//add_action( 'login_form', array(&$this, 'wmcaptcha_display'),100,1);
					//add_filter( 'authenticate', array(&$this, 'registration_captcha_errors'),22,1);	
	
				}
			}
			if($wmsc_options['captcha_enable_registration']){
				add_action( 'init', 	array(&$this, 'wmsimplecaptcha_init_enqueue_scripts'));
				add_action( 'wp_ajax_wmsimplecaptcha_action', 			array(&$this, 'wmsimplecaptcha_action'));
				add_action( 'wp_ajax_nopriv_wmsimplecaptcha_action', 	array(&$this, 'wmsimplecaptcha_action'));
			}
		}// End Function __construct()
		
		public function include_admin() {
			include_once( 'wm_simple_captcha_admin.php' );
		}
		function define_constant($filename){
			
			$uploads = $this->font_upload_dir('dir');
			
			if(!defined('WM_SIMPLE_CAPTCHA_FILE_PATH')) 	define('WM_SIMPLE_CAPTCHA_FILE_PATH', 	dirname( $filename ) );
			if(!defined('WM_SIMPLE_CAPTCHA_DIR_NAME')) 		define('WM_SIMPLE_CAPTCHA_DIR_NAME', 	basename( WM_SIMPLE_CAPTCHA_FILE_PATH ) );
			if(!defined('WM_SIMPLE_CAPTCHA_FOLDER')) 		define('WM_SIMPLE_CAPTCHA_FOLDER', 		dirname( plugin_basename( $filename ) ) );
			if(!defined('WM_SIMPLE_CAPTCHA_NAME')) 			define('WM_SIMPLE_CAPTCHA_NAME', 		plugin_basename($filename) );
						
			if(!defined('WM_SIMPLE_CAPTCHA_URL')) 			define('WM_SIMPLE_CAPTCHA_URL', 		WP_PLUGIN_URL ."/". WM_SIMPLE_CAPTCHA_DIR_NAME );
			if(!defined('WM_SIMPLE_CAPTCHA_CODE_URL')) 		define('WM_SIMPLE_CAPTCHA_CODE_URL', 	WM_SIMPLE_CAPTCHA_URL ."/captcha_code/captcha_code.php" );	
			if(!defined('WM_SIMPLE_CAPTCHA_FONT_PATH')) 	define('WM_SIMPLE_CAPTCHA_FONT_PATH',	WM_SIMPLE_CAPTCHA_FILE_PATH . '/fonts/');
			
			if(!defined('WM_SIMPLE_CAPTCHA_NEW_FONT_PATH')) 	define('WM_SIMPLE_CAPTCHA_NEW_FONT_PATH',	$uploads  . '/wm_simple_captcha_fonts/');
			
			
		}// End Function define_constant()
		
		function registration_captcha_errors( $errors = NULL ) {
			global $wmsc_options;
			
			$err = $this->captcha_errors();			
			if($err){
				//if($errors == NULL) global $errors;
				if($errors == NULL) $errors = new WP_Error();
				$errors = new WP_Error();
				if($err == "empty"){
					$errors->add( 'recaptcha', __( str_replace("ERROR:","<strong>ERROR: </strong>", $wmsc_options['captcha_empty']), 'wmsimplecaptcha' ), 'invalid-site-private-key' );
				}
				
				if($err == "invalid"){
					$errors->add( 'recaptcha', __( str_replace("ERROR:","<strong>ERROR: </strong>", $wmsc_options['captcha_invalid']), 'wmsimplecaptcha' ), 'invalid-site-private-key' );
				}				
			}
			
			return $errors;
		}// End Function registration_captcha_errors()
		
		function login_captcha_errors( $errors_str = NULL ) {
			global $error;
			$err = $this->captcha_errors();
			if($err){
				if($errors_str == NULL) $errors_str = '';
				$errors_str = '';
				if($err == "empty"){
					$errors_str .= $wmsc_options['captcha_empty'];
				}
				
				if($err == "invalid"){
					$errors_str .= $wmsc_options['captcha_invalid'];	
				}				
			}
			
			
			return $errors_str;
		}// End Function login_captcha_errors()
		
		function captcha_errors($errors = NULL){
			if(isset($_REQUEST['captcha_challenge_field'])){
				// session_start();
				$captcha_challenge_field = strtolower(trim($_REQUEST['captcha_challenge_field']));
			
				if(strlen($captcha_challenge_field)<=0){
					$errors = "empty";
				}
				
				if(strlen($captcha_challenge_field)>0){
					if(isset($_SESSION['6_letters_code'])){
						$code = strtolower($_SESSION['6_letters_code']);
						if($code != $captcha_challenge_field){							
							$errors = "invalid";
						}
					}					
				}
			}
			return $errors;
		}// End Function captcha_errors()
		
		function wmcaptcha_display(){
			
			$new_option 				= $this->wmcaptcha_form_values();
			$captcha_custom_template	= apply_filters('wmsimplecaptcha_captcha_custom_template',NULL, $new_option);
			
			if($captcha_custom_template){
				echo $captcha_custom_template;
			}else{
				?>
				<p class="wmcaptcha_box">
					<?php if($new_option['captcha_label']):?>
					<label for="captcha_challenge_field"><?php _e( $new_option['captcha_label'], 'wmsimplecaptcha' ); ?></label>
					<?php endif;?>
					<span class="wmcaptcha_fieldbox">
						<span class="wmcaptcha_fieldbox_input"><input type="text" name="captcha_challenge_field" id="captcha_challenge_field" class="captcha_challenge_field" maxlength="<?php echo $new_option['captcha_image_characters'];?>" /></span>
						<span class="wmcaptcha_fieldbox_img <?php echo $new_option['captcha_refresh_image_class'];?>"><img src="<?php echo $new_option['captcha_url'];?>" class="captcha_code_img" id="captcha_code_file" alt="" /></span>
						<?php if($new_option['captcha_enable_refresh']):?>
							<span class="wmcaptcha_fieldbox_a refresh_button"><?php echo $new_option['refresh_image'];?></span>
						<?php endif;?>
						<span class="wmcaptcha_fieldbox_clearfix clearfix"></span>
					</span>
					<span class="wmcaptcha_clearfix"></span>
				</p>
				<?php
				//echo $new_option['captcha_url'];
			}			
			add_action('wp_footer', 				array(&$this, 'wmsimplecaptcha_wp_footer'));
			add_action('login_footer', 				array(&$this, 'wmsimplecaptcha_wp_footer'));			
		}// End Function wmcaptcha_display()
		
		function wmcaptcha_form_values($errors = NULL ){
			global $wmsc_options, $new_option;
			if($errors == NULL) $errors = new WP_Error();
					
			$captcha_enable_css 			= $wmsc_options['captcha_enable_css'];
			$captcha_image_characters 		= ($wmsc_options['captcha_image_characters'] + apply_filters( 'wmsimplecaptcha_extra_maxlength',0));
			$captcha_enable_refresh 		= $this->isset_option('captcha_enable_refresh',$wmsc_options,0);
			$captcha_enable_refresh_image 	= $this->isset_option('captcha_enable_refresh_image',$wmsc_options,0);			
			$captcha_label 					= $this->isset_option('captcha_label',$wmsc_options,NULL);
			$captcha_refresh_image 			= $wmsc_options['captcha_refresh_image'];
			$captcha_refresh_image 			= apply_filters( 'wmsimplecaptcha_captcha_refresh_image',$captcha_refresh_image);			
			$captcha_refresh_text 			= apply_filters( 'wmsimplecaptcha_captcha_refresh_text','Refresh');
			$captcha_enable_refresh_text	= apply_filters( 'wmsimplecaptcha_captcha_enable_refresh_text',false);
			$captcha_url					= WM_SIMPLE_CAPTCHA_CODE_URL.'?captcha_code='.rand(1111,9999);
			
			$refresh_image = NULL;
			if($captcha_refresh_image && $captcha_enable_refresh_text == false){				
				$refresh_image = '<img src="'.$captcha_refresh_image.'" alt="'.$captcha_refresh_text.'" />';
			}else{
				$refresh_image = $captcha_refresh_text;
			}
			$captcha_refresh_image_class = "";
			if($captcha_enable_refresh_image) $captcha_refresh_image_class = " refresh_image";
			
			
			$new_option = array();
			$new_option['captcha_image_characters'] 	= $captcha_image_characters;
			$new_option['refresh_button_url'] 			= $captcha_refresh_image;
			$new_option['captcha_enable_css'] 			= $captcha_enable_css;
			$new_option['captcha_enable_refresh'] 		= $captcha_enable_refresh;
			$new_option['captcha_enable_refresh_image'] = $captcha_enable_refresh_image;
			$new_option['refresh_image'] 				= $refresh_image;
			$new_option['captcha_label'] 				= $captcha_label;
			$new_option['captcha_refresh_text']			= $captcha_refresh_text;
			$new_option['captcha_enable_refresh_text']	= $captcha_enable_refresh_text;
			$new_option['captcha_url']					= $captcha_url;
			$new_option['captcha_refresh_image_class']	= $captcha_refresh_image_class;
			return $new_option;
		}// End Function wmcaptcha_form_values()
		
		function wmsimplecaptcha_scripts_front() {
			global $wmsc_options;
			$captcha_enable_css 		= $wmsc_options['captcha_enable_css'];
			if($captcha_enable_css)
			wp_enqueue_style('wmsimplecaptcha_style_front',WM_SIMPLE_CAPTCHA_URL . '/assets/css/wmsimplecaptcha_style.css');			
		}	
		function wmsimplecaptcha_init_enqueue_scripts($hook) {
			if( 'index.php' != $hook ) {
				//return;
			}				
			wp_enqueue_script( 'wmsimplecaptcha_scripts_front', WM_SIMPLE_CAPTCHA_URL.'/assets/js/wmsimplecaptcha_scripts.js', array('jquery') );		
			wp_localize_script('wmsimplecaptcha_scripts_front', 'ajax_object',	array( 'ajax_url' => admin_url( 'admin-ajax.php' ) ) );
		}
		
		
		function wmsimplecaptcha_action() {
			$captcha_code = rand(111,9999);
			echo '<img src="'.WM_SIMPLE_CAPTCHA_CODE_URL.'?captcha_code='.$captcha_code.'" class="captcha_code_img" id="captcha_code_file"  />';
			die();
		}
		
		function wmsimplecaptcha_wp_footer() {
			global $wmsc_options;
			
			$captcha_enable_css 		= $wmsc_options['captcha_enable_css'];
			$captcha_background_color 	= $this->isset_option('captcha_background_color',$wmsc_options,NULL);
			$captcha_image_height 		= $this->isset_option('captcha_image_height',$wmsc_options,NULL);
			$captcha_image_width 		= $this->isset_option('captcha_image_width',$wmsc_options,NULL);
			
			$captcha_enable_border 		= $this->isset_option('captcha_enable_border',$wmsc_options,NULL);
			$captcha_border_width 		= $this->isset_option('captcha_border_width',$wmsc_options,0);
			$captcha_border_type 		= $this->isset_option('captcha_border_type',$wmsc_options,NULL);
			$captcha_border_color 		= $this->isset_option('captcha_border_color',$wmsc_options,NULL);
			
			$captcha_custom_css 		= $this->isset_option('captcha_custom_css',$wmsc_options,NULL);
			
			$output = "";
			if($captcha_enable_css || $captcha_enable_border || $captcha_custom_css){
				$output .= '<style type="text/css">';
				
				if($captcha_enable_css == 1){
					if(!$captcha_enable_border){
						$captcha_border_width = 0;
					}
					$output .= '.wmcaptcha_box .wmcaptcha_fieldbox_img{
						width:'.($captcha_image_width+$captcha_border_width).'px;
						height:'.($captcha_image_height+$captcha_border_width).'px;
						background-color:'.$captcha_background_color.';
					}';
				}
				
				if($captcha_enable_border == 1){
					$output .= '.wmcaptcha_box .wmcaptcha_fieldbox_img img{				
						border-width:'.$captcha_border_width.'px;
						border-style:'.$captcha_border_type.';
						border-color:'.$captcha_border_color.';
					}';
				}
				$output .= $captcha_custom_css;
				$output .= '</style>';
			}
    		echo $output;
		}
		
		
		
		function isset_option($name = NULL,$data = NULL, $detault = ""){
			$r = $detault;
			if($data && isset($data[$name])){
				$r = $data[$name];
			}
			
			return $r;
		}// End Function isset_option()
		
		
		function print_array($ar = NULL,$display = true){
			if($ar){
				$output = "<pre>";
				$output .= print_r($ar,true);
				$output .= "</pre>";
				
				if($display){
					echo $output;
				}else{
					return $output;
				}
			}
		}// End Function print_array()
		
		function default_values(){
				$default = array();	
				$default["captcha_enable_registration"] 	= "0";
				$default["captcha_image_width"] 			= "120";
				$default["captcha_image_height"] 			= "40";
				$default["captcha_image_characters"] 		= "4";
				$default["captcha_image_font_adj"]			= "0.6";
				$default["captcha_enable_space"]			= "0";
				$default["captcha_image_font"] 				= "arial.ttf";
				$default["captcha_possible_letters"] 		= "23456789";
				$default["captcha_random_dots"] 			= "392";
				$default["captcha_random_lines"] 			= "286";
				$default["captcha_text_color"] 				= "#ffffff";
				$default["captcha_dots_color"] 				= "#27d141";
				$default["captcha_line_color"]				= "#ff2d2d";
				$default["captcha_background_color"]		= "#f4f4f4";
				$default["captcha_label"] 					= "Security Code";
				$default["captcha_enable_css"] 				= "0";
				$default["captcha_enable_border"] 			= "0";
				$default["captcha_border_width"] 			= "1";
				$default["captcha_border_type"]				= "solid";
				$default["captcha_border_color"] 			= "#c4c4c4";
				$default["captcha_enable_refresh_image"] 	= "0";
				$default["captcha_enable_refresh"] 			= "0";
				$default["captcha_refresh_image"] 			= "";
				$default["captcha_empty"] 					= "ERROR: Please enter security code.";
				$default["captcha_invalid"] 				= "ERROR: Please enter valid security code.";
				$default["captcha_custom_css"] 				= "";
				$default["font_path"] 						= WM_SIMPLE_CAPTCHA_NEW_FONT_PATH;
				
				
				$wmsc_options = get_option('wmsimplecaptcha');
				
				if(!$wmsc_options){
					$default = $this->activate();
					add_option('wmsimplecaptcha', $default);
				}				
				$wmsc_options 	= array_merge((array)$default, (array)$wmsc_options);				
				return $wmsc_options;
		}// End Function define_constant()
		
		function activate() {
				$default = array();
				$default["captcha_enable_registration"] 	= "1";
				$default["captcha_image_width"] 			= "120";
				$default["captcha_image_height"] 			= "40";
				$default["captcha_image_characters"] 		= "4";
				$default["captcha_image_font_adj"]			= "0.6";
				$default["captcha_enable_space"]			= "0";
				$default["captcha_image_font"] 				= "arial.ttf";
				$default["captcha_possible_letters"] 		= "23456789";
				$default["captcha_random_dots"] 			= "392";
				$default["captcha_random_lines"] 			= "286";
				$default["captcha_text_color"] 				= "#ffffff";
				$default["captcha_dots_color"] 				= "#27d141";
				$default["captcha_line_color"]				= "#ff2d2d";
				$default["captcha_background_color"]		= "#f4f4f4";
				$default["captcha_label"] 					= "Security Code";
				$default["captcha_enable_css"] 				= "1";
				$default["captcha_enable_border"] 			= "0";
				$default["captcha_border_width"] 			= "1";
				$default["captcha_border_type"]				= "solid";
				$default["captcha_border_color"] 			= "#c4c4c4";
				$default["captcha_enable_refresh_image"] 	= "0";
				$default["captcha_enable_refresh"] 			= "0";
				$default["captcha_refresh_image"] 			= "";
				$default["captcha_empty"] 					= "ERROR: Please enter security code.";
				$default["captcha_invalid"] 				= "ERROR: Please enter valid security code.";
				$default["captcha_custom_css"] 				= "";			
				add_option( 'wmsimplecaptcha', $default );
				return $default;
		}
		
		/**
		 * Deactivate
		 * @return boolean
		 */
		function deactivate() {
			delete_option('wmsimplecaptcha_activated_plugin_error');
		}
		
		/**
		 * Tidy up deleted plugin by removing options
		 */
		static function uninstall() {		
			delete_option('wmsimplecaptcha');
			delete_option('wmsimplecaptcha_activated_plugin_error');			
		}
		
		function font_upload_dir( $type = false ) {
			$uploads = wp_upload_dir();
		
			$uploads = apply_filters( 'wm-simple-captcha-dir', array(
				'dir' => $uploads['basedir'],
				'url' => $uploads['baseurl'] ) );
		
			if ( 'dir' == $type ){
				
				
				return $uploads['dir'];
			}if ( 'url' == $type )
				return $uploads['url'];
				
			
		
			return $uploads;
		}
	}// End Class WM_Simple_Captcha_Front()
}